python/ray/tune/trial_runner.py

```diff
@@ -322,14 +323,33 @@ def __init__(self,
 
         self._callbacks = CallbackList(callbacks or [])
 
-        self._callbacks.setup()
-
         if checkpoint_period is None:
             checkpoint_period = os.getenv("TUNE_GLOBAL_CHECKPOINT_S", "auto")
 
         self._checkpoint_period = checkpoint_period
         self._checkpoint_manager = self._create_checkpoint_manager()
 
+    def setup_experiments(self, experiments: List[Experiment],
```

Comment:
This is not "setting up" experiment right? This is sort of on_experiment/searcher_setup right?

Also I am wondering how's information flowing? Seems like
Tune config --> Experiment --> Searcher --> Callback
                                       \---> Callback
You sort of captured the flow of information to Callback in this function here. Ideally I would want to reuse this function for flow of information between Experiment and Searcher. Things would be much easier if every information here is coming from Experiment to some other components (whether Searcher or Callback).

Reply:
It was `setup_callbacks`, but we've changed it to be more generic. Though you are right, this may still not be the best name.

The issue with using this for passing to searcher/scheduler is that `set_search_properties` on those is called before this (and thus would miss information like `total_samples`). Perhaps we could add a new method to searchers/schedulers instead of extending `set_search_properties`, so that we can use it here. @krfricke What do you think?

Reply:
Can `total_samples` be generated earlier?

Reply:
We need the config spaces for it so it has to happen *after* `set_search_properties`, as that sets the config